javascript/ql/src/semmle/javascript/dataflow/Nodes.qll

```diff
@@ -462,3 +462,255 @@ DataFlow::SourceNode moduleMember(string path, string m) {
     result = DataFlow::ssaDefinitionNode(ssa)
   )
 }
+
+/**
+ * A data flow node corresponding to a class definition or a function definition
+ * acting as a class.
+ *
+ * The following patterns are recognized as classes and methods:
+ * ```
+ * class C {
+ *   method()
+ * }
+ *
+ * function F() {}
+ *
+ * F.prototype.method = function() {}
+ *
+ * F.prototype = {
+ *   method: function() {}
+ * }
+ *
+ * extend(F.prototype, {
+ *   method: function() {}
+ * });
+ * ```
+ *
+ * Additional patterns can be recognized as class nodes, by extending `DataFlow::ClassNode::Range`.
+ */
+class ClassNode extends DataFlow::SourceNode {
+  ClassNode::Range impl;
+
+  ClassNode() { this = impl }
+
+  /**
+   * Gets the name of the class, if it has one.
+   */
+  string getName() { result = impl.getName() }
+
+  /**
+   * Gets a description of the class.
+   */
+  string desribe() { result = impl.describe() }
+
+  /**
+   * Gets the constructor function of this class.
+   */
+  FunctionNode getConstructor() { result = impl.getConstructor() }
+
+  /**
+   * Gets an instance method declared in this class, with the given name, if any.
+   *
+   * Does not include methods from superclasses.
+   */
+  FunctionNode getAnInstanceMethod(string name) { result = impl.getAnInstanceMethod(name) }
```

Comment:
Here and below, I wonder whether it is really worth including "a"/"an" in the predicate name and doc comment. Of course, it is technically correct since we don't assume/enforce uniqueness, but in the majority of cases we probably would expect the result to be unique, so having an overly general name might lead to confusion and bad auto-complete experience.

What do you think?

(We should, of course, still mention the possibility of multiple results in the doc comment.)

Reply:
I considered this as well, for the same reasons, but I think "a"/"an" should be used here.

Some arguments for:

- the rest of the `SourceNode` getters use that convention
- multiple prototypes are possible
- the method definitions may be conditional
- unlike ES6 classes, these class nodes are not always syntactic, the "a"/"an" prefix kind of reflects that
- if we later add support for lookups in the prototype hierarchy, I think we want the "a/an" prefix, so it would be nice if we had some consistency between similar lookups

(this list is probably a rehash of what you already know)

Reply:
I'm also against the a/an prefix. Even if it's technically possible, it's confusing and extremely rare in practice, and inconsistent with other APIs for things that are unique in practice.

Reply:
Ok. Let us remove it then.